src/core/server/http/http_server.ts

```diff
@@ -167,66 +167,49 @@ export class HttpServer {
           xsrfRequired: route.options.xsrfRequired ?? !isSafeMethod(route.method),
         };
 
-        // To work around https://github.com/hapijs/hapi/issues/4122 until v20, set the socket
-        // timeout on the route to a fake timeout only when the payload timeout is specified.
-        // Within the onPreAuth lifecycle of the route itself, we'll override the timeout with the
-        // real socket timeout.
-        const fakeSocketTimeout = timeout?.payload ? timeout.payload + 1 : undefined;
-
-        this.server.route({
+        const routeOpts: ServerRoute = {
           handler: route.handler,
           method: route.method,
           path: route.path,
           options: {
             auth: this.getAuthOption(authRequired),
             app: kibanaRouteOptions,
-            ext: {
-              onPreAuth: {
-                method: (request, h) => {
-                  // At this point, the socket timeout has only been set to work-around the HapiJS bug.
-                  // We need to either set the real per-route timeout or use the default idle socket timeout
-                  if (timeout?.idleSocket) {
-                    request.raw.req.socket.setTimeout(timeout.idleSocket);
-                  } else if (fakeSocketTimeout) {
-                    // NodeJS uses a socket timeout of `0` to denote "no timeout"
-                    request.raw.req.socket.setTimeout(this.config!.socketTimeout ?? 0);
-                  }
-
-                  return h.continue;
-                },
-              },
-            },
             tags: tags ? Array.from(tags) : undefined,
-            // TODO: This 'validate' section can be removed once the legacy platform is completely removed.
-            // We are telling Hapi that NP routes can accept any payload, so that it can bypass the default
-            // validation applied in ./http_tools#getServerOptions
-            // (All NP routes are already required to specify their own validation in order to access the payload)
-            validate,
-            payload: [allow, maxBytes, output, parse, timeout?.payload].some(
-              (v) => typeof v !== 'undefined'
-            )
+            // @ts-expect-error Types are outdated and doesn't allow `payload.multipart` to be `true`
+            payload: [allow, maxBytes, output, parse, timeout?.payload].some((x) => x !== undefined)
               ? {
                   allow,
                   maxBytes,
                   output,
                   parse,
                   timeout: timeout?.payload,
+                  multipart: true, // TODO: Figure out what the default value should be
                 }
               : undefined,
             timeout: {
-              socket: fakeSocketTimeout,
+              socket: timeout?.idleSocket ?? this.config!.socketTimeout ?? false, // TODO: Do we really want to fall back to NO timeout? Shouldn't we default to the default Node.js timeout of 2 minutes?
```

Comment:
@restrry I had to fall back to `false` instead of `0` because hapi is validating the value and doesn't allow zero. I'm not sure if `0` previously meant "no timeout", but now `false` means "no timeout", so I think this is what the old code intended.

However, I do think that having no timeout at all is not good practise. But I don't know if we did that on purpose?

Reply:
> I do think that having no timeout at all is not good practise.

It seems to be a workaround for introduced by @kobelb in https://github.com/elastic/kibana/pull/73730

https://github.com/elastic/kibana/blob/e116fce08eda66130a2169ece880e26b33cf25e0/src/core/server/http/http_server.ts#L200-L201

As this problem is fixed in hapi v20 it should be okay to fallback to `undefined` @kobelb could you confirm, please?

Reply:
Thanks for looking into this Mikhail ☺️

I've updated the code to not fall back to `false`, which should mean we just fall back to `undefined` as that should be the value of `this.config.socketTimeout` if it's not set.

We'll see if we can keep it like this based on @kobelb's feedback.

Reply:
> I've updated the code to not fall back to false, which should mean we just fall back to undefined as that should be the value of this.config.socketTimeout if it's not set.

LGTM <3